File: cpp/2799-Count Complete Subarrays in an Array.cpp
#include <vector>
#include <iostream>
#include <set>
#include <unordered_map>

class Solution
{
public:
  int countCompleteSubarrays(vector<int>& nums)
  {
    // Early out
    if(nums.size() == 1)
    {
      return 1;
    }

    std::set<int> uniqueElems;
    for(const int val : nums)
    {
      uniqueElems.insert(val);
    }
    int numUniqueElems = uniqueElems.size();

    int ret = 0;
    int leftPtr = 0;
    while(leftPtr <= nums.size() - numUniqueElems)
    {
      std::unordered_map<int, int> uniqueValuesToCounts;
      int rightPtr = leftPtr;
      while(rightPtr < nums.size())
      {
        //std::cout << "leftPtr " << leftPtr << " rightPtr " << rightPtr << "\n";
        uniqueValuesToCounts[nums[rightPtr]]++;
        if(uniqueValuesToCounts.size() == numUniqueElems)
        {
          ret++;
          rightPtr++;
        }
        else if(uniqueValuesToCounts.size() < numUniqueElems)
        {
          rightPtr++;
        }
        else // uniqueValuesToCounts.size() > numUniqueElems
        {
          break;
        }
      }
      leftPtr++;
    }

    return ret;
  }
};

File: python/2799-Count Complete Subarrays in an Array.py
class Solution(object):
  def countCompleteSubarrays(self, nums):
    """
    :type nums: List[int]
    :rtype: int
    """
    uniqueElems = {}
    for n in nums:
      uniqueElems[n] = 1
    
    left = 0
    ret = 0
    n = len(nums)
    while left < n - len(uniqueElems) + 1:
      right = left
      thisElem = {}
      while right < n:
        if nums[right] in thisElem:
          thisElem[nums[right]] += 1
        else:
          thisElem[nums[right]] = 1
        if len(uniqueElems) == len(thisElem):
          ret += 1
        right += 1
      left += 1
    return ret